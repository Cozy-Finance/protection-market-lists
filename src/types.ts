export interface Version {
  readonly major: number;
  readonly minor: number;
  readonly patch: number;
}

export interface Tags {
  readonly [tagId: string]: {
    readonly name: string;
    readonly description: string;
  };
}

export interface InvestmentOpportunity {
  readonly divest: DivestScript;
  readonly dataURI: string; // API endpoint to query for user data from, e.g. balances
  readonly invest: InvestScript;
  readonly logoURI?: string;
  readonly strategy?: StrategyLeg[];
  readonly tokens: {
    // arrays of token addresses, to be looked up in the `tokens` mapping
    readonly input: string[]; // token addresses the user will deposit to enter this opportunity, and receipt upon exiting
    readonly receipt: string[]; // receipt token addresses the user needs to exit this opportunity
    readonly rewards: string[]; // rewards token addresses the user may claim when exiting this opportunity
  };
}

// Parameters for investing into a protected investment.
//
// Example  'function invest(uint256[] calldata amounts) external payable'
//
//  The `tokens` property defines the  set of input tokens that can be used to deposit into this investment opportunity.
// The items in the function signature correspond to these `tokens`. For example, if we're investing in the 3Crypto pool,
// the `tokens` array would be [ETH, WBTC, USDT]. When calling `invest` on `address`, amounts[0] would correspond to the
// amount of ETH to invest, amounts[1] would correspond to the amount of WBTC to invest, etc. An amount of MAX_UINT256
// should be used to signal that you want to invest all tokens you have
export interface InvestScript {
  readonly address: string;
  readonly signature: string;
}

// Parameters for exiting a protected investment.
//
// Function signature meaning is analogous to `InvestScript`. The `amounts` and `claimRewards` parameters are uint256
// where zero means don't withdraw/claim anything and MAX_UINT256 means withdraw/claim all. Note that not all
// reward tokens will support arbitrary claim amounts. The `recipient` field specifies the address all output
// tokens and reward tokens should be sent to
export interface DivestScript {
  readonly address: string;
  readonly signature: string;
}

export interface StrategyLeg {
  readonly action: 'Borrow' | 'Swap to' | 'Deposit for' | 'Deposit to' | 'Stake on';
  readonly actionReceiver: string;
  readonly logoURI?: string;
  readonly platformId: number; // Platform ID of strategy target
  readonly token: string; // address of the token, to be looked up in the `tokens` mapping
}

export interface MarketList {
  readonly keywords?: string[];
  readonly logoURI?: string;
  readonly markets: MarketInfo[];
  readonly name: string;
  readonly tags?: Tags;
  readonly timestamp: string;
  readonly tokens: Record<string, TokenInfo>; // maps from checksummed token address to TokenInfo
  readonly version: Version;
}

export interface MarketInfo extends TokenInfo {
  readonly investmentOpportunity?: InvestmentOpportunity; // details on how protected investing using this market
  readonly trigger: string; // trigger contract address for protection markets, or zero address for money markets
  readonly underlying: string; // address of underlying, if `underlying === 0xEeee....EEeE`, underlying is ETH
}

export interface TokenInfo {
  readonly address: string;
  readonly chainId: number;
  readonly decimals: number;
  readonly extensions?: {
    readonly [key: string]: string | number | boolean | null;
  };
  readonly logoURI?: string;
  readonly name: string;
  readonly symbol: string;
  readonly tags?: string[];
}
