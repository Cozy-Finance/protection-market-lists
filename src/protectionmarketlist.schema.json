{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Protection Market List",
  "description": "Schema for lists of markets available on the Cozy Protocol",
  "definitions": {
    "Version": {
      "type": "object",
      "description": "The version of the list, used in change detection",
      "examples": [
        {
          "major": 1,
          "minor": 0,
          "patch": 0
        }
      ],
      "additionalProperties": false,
      "properties": {
        "major": {
          "type": "integer",
          "description": "The major version of the list. Must be incremented when tokens are removed from the list or token addresses are changed.",
          "minimum": 0,
          "examples": [1, 2]
        },
        "minor": {
          "type": "integer",
          "description": "The minor version of the list. Must be incremented when tokens are added to the list.",
          "minimum": 0,
          "examples": [0, 1]
        },
        "patch": {
          "type": "integer",
          "description": "The patch version of the list. Must be incremented for any changes to the list.",
          "minimum": 0,
          "examples": [0, 1]
        }
      },
      "required": ["major", "minor", "patch"]
    },
    "TagIdentifier": {
      "type": "string",
      "description": "The unique identifier of a tag",
      "minLength": 1,
      "maxLength": 30,
      "pattern": "^[\\w ]+$",
      "examples": ["yearn", "stablecoin", "money market", "protection market"]
    },
    "ExtensionIdentifier": {
      "type": "string",
      "description": "The name of a market extension property",
      "minLength": 1,
      "maxLength": 30,
      "pattern": "^[\\w ]+$",
      "examples": ["is_audited", "is_triggered"]
    },
    "ExtensionValue": {
      "anyOf": [
        {
          "type": "string",
          "minLength": 1,
          "maxLength": 42,
          "examples": ["#00000"]
        },
        {
          "type": "boolean",
          "examples": [true]
        },
        {
          "type": "number",
          "examples": [15]
        },
        {
          "type": "null"
        }
      ]
    },
    "TagDefinition": {
      "type": "object",
      "description": "Definition of a tag that can be associated with a market via its identifier",
      "additionalProperties": false,
      "properties": {
        "name": {
          "type": "string",
          "description": "The name of the tag",
          "pattern": "^[ \\w]+$",
          "minLength": 1,
          "maxLength": 20
        },
        "description": {
          "type": "string",
          "description": "A user-friendly description of the tag",
          "pattern": "^[ \\w\\.,:]+$",
          "minLength": 1,
          "maxLength": 200
        }
      },
      "required": ["name", "description"],
      "examples": [
        {
          "name": "Stablecoin",
          "description": "A market that takes stablecoin deposits"
        }
      ]
    },
    "DivestScript": {
      "type": "object",
      "description": "Parameters of the contract containing a `divest` method to exit an investment opportunity",
      "properties": {
        "address": {
          "type": "string"
        },
        "signature": {
          "enum": [
            "function divest(address recipient, uint256[] receiptAmounts, uint256[] rewardAmounts)"
          ],
          "type": "string"
        }
      }
    },
    "InvestScript": {
      "type": "object",
      "description": "Parameters of the contract containing an `invest` method to enter an investment opportunity",

      "properties": {
        "address": {
          "type": "string"
        },
        "signature": {
          "enum": ["function invest(uint256[] amounts, uint256[] approve)"],
          "type": "string"
        }
      }
    },
    "InvestmentOpportunity": {
      "type": "object",
      "description": "Properties required for detailing how a user can enter a protected investment after borrowing from the protection market",
      "properties": {
        "dataURI": {
          "type": "string",
          "description": "API endpoint that can be queried to return market and user data about the investment opportunity",
          "format": "uri",
          "examples": [
            "https://api.cozy.finance/opportunities/usdc-yearn?account=0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045"
          ]
        },
        "divest": {
          "$ref": "#/definitions/DivestScript"
        },
        "invest": {
          "$ref": "#/definitions/InvestScript"
        },
        "strategy": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/StrategyLeg"
          }
        },
        "tokens": {
          "type": "object",
          "properties": {
            "input": {
              "type": "array",
              "description": "Array of one or more tokens a user must have to invest in this opportunity",
              "items": {
                "$ref": "#/definitions/TokenInfo"
              }
            },
            "receipt": {
              "type": "array",
              "description": "Array of one or more tokens a user will receive after investing in this opportunity",
              "items": {
                "$ref": "#/definitions/TokenInfo"
              }
            },
            "rewards": {
              "type": "array",
              "description": "Array of one or more tokens a user will receive as a reward for investing in this opportunity",
              "items": {
                "$ref": "#/definitions/TokenInfo"
              }
            }
          }
        }
      }
    },
    "ProtectionMarketInfo": {
      "type": "object",
      "description": "Metadata for a single market in a protection market list",
      "additionalProperties": false,
      "properties": {
        "chainId": {
          "type": "integer",
          "description": "The chain ID of the Ethereum network where this token is deployed",
          "minimum": 1,
          "examples": [1, 42]
        },
        "address": {
          "type": "string",
          "description": "The checksummed address of the protection market on the specified chain ID",
          "pattern": "^0x[a-fA-F0-9]{40}$",
          "examples": ["0xdBDF2fC3Af896e18f2A9DC58883d12484202b57E"]
        },
        "decimals": {
          "type": "integer",
          "description": "The number of decimals the market's receipt token has",
          "minimum": 0,
          "maximum": 255,
          "examples": [18]
        },
        "name": {
          "type": "string",
          "description": "The name of the protection market",
          "minLength": 1,
          "maxLength": 100,
          "pattern": "^[ \\w.'+\\-%\\/À-ÖØ-öø-ÿ:]+$",
          "examples": ["Cozy USD Coin"]
        },
        "symbol": {
          "type": "string",
          "description": "The symbol for the protection market; must be alphanumeric",
          "pattern": "^[a-zA-Z0-9+\\-%\\/\\$]+$",
          "minLength": 1,
          "maxLength": 100,
          "examples": ["cozyUSDC"]
        },
        "logoURI": {
          "type": "string",
          "description": "A URI to the protection market's logo asset; if not set, interface will attempt to find a logo based on the token address; suggest SVG or PNG of size 256x256",
          "format": "uri",
          "examples": ["ipfs://QmXfzKRvjZz3u5JRgC4v5mGVbm9ahrUiB4DgzHBsnWbTMM"]
        },
        "investmentOpportunity": {
          "$ref": "#/definitions/InvestmentOpportunity"
        },
        "isTriggered": {
          "type": "boolean",
          "description": "True if this market has been triggered, false if not triggered or not applicable",
          "examples": [true, false]
        },
        "trigger": {
          "type": "string",
          "description": "The checksummed address of the market's trigger contract, or the zero address for markets without a trigger",
          "pattern": "^0x[a-fA-F0-9]{40}$",
          "examples": [
            "0x02E57c5800a345Fe97d73fA8bB72B948E19aA499",
            "0x0000000000000000000000000000000000000000"
          ]
        },
        "underlying": {
          "$ref": "#/definitions/TokenInfo"
        },
        "tags": {
          "type": "array",
          "description": "An array of tag identifiers associated with the market; tags are defined at the list level",
          "items": {
            "$ref": "#/definitions/TagIdentifier"
          },
          "maxItems": 10,
          "examples": ["stablecoin", "compound"]
        },
        "extensions": {
          "type": "object",
          "description": "An object containing any arbitrary or vendor-specific market metadata",
          "propertyNames": {
            "$ref": "#/definitions/ExtensionIdentifier"
          },
          "additionalProperties": {
            "$ref": "#/definitions/ExtensionValue"
          },
          "maxProperties": 10,
          "examples": [
            {
              "color": "#000000",
              "is_verified_by_me": true
            }
          ]
        }
      }
    },
    "StrategyLeg": {
      "type": "object",
      "description": "Describes one of the steps taken to enter an investment opportunity",
      "properties": {
        "action": {
          "enum": [
            "Borrow",
            "Swap to",
            "Deposit for",
            "Deposit to",
            "Stake on"
          ],
          "type": "string"
        },
        "destination": {
          "type": "number",
          "description": "Platform ID the strategy leg targets",
          "minimum": 0
        },
        "logoURI": {
          "type": "string",
          "description": "A URI for the logo of the strategy leg; prefer SVG or PNG of size 256x256",
          "format": "uri",
          "examples": ["ipfs://QmXfzKRvjZz3u5JRgC4v5mGVbm9ahrUiB4DgzHBsnWbTMM"]
        }
      }
    },
    "TokenInfo": {
      "type": "object",
      "description": "Metadata for a single token",
      "additionalProperties": false,
      "properties": {
        "chainId": {
          "type": "integer",
          "description": "The chain ID of the Ethereum network where this token is deployed",
          "minimum": 1,
          "examples": [1, 42]
        },
        "address": {
          "type": "string",
          "description": "The checksummed address of the token on the specified chain ID",
          "pattern": "^0x[a-fA-F0-9]{40}$",
          "examples": ["0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"]
        },
        "decimals": {
          "type": "integer",
          "description": "The number of decimals for the token balance",
          "minimum": 0,
          "maximum": 255,
          "examples": [18]
        },
        "name": {
          "type": "string",
          "description": "The name of the token",
          "minLength": 1,
          "maxLength": 40,
          "pattern": "^[ \\w.'+\\-%\\/À-ÖØ-öø-ÿ:]+$",
          "examples": ["USD Coin"]
        },
        "symbol": {
          "type": "string",
          "description": "The symbol for the token; must be alphanumeric",
          "pattern": "^[a-zA-Z0-9+\\-%\\/\\$]+$",
          "minLength": 1,
          "maxLength": 20,
          "examples": ["USDC"]
        },
        "logoURI": {
          "type": "string",
          "description": "A URI to the token logo asset; if not set, interface will attempt to find a logo based on the token address; suggest SVG or PNG of size 256x256",
          "format": "uri",
          "examples": ["ipfs://QmXfzKRvjZz3u5JRgC4v5mGVbm9ahrUiB4DgzHBsnWbTMM"]
        },
        "tags": {
          "type": "array",
          "description": "An array of tag identifiers associated with the token; tags are defined at the list level",
          "items": {
            "$ref": "#/definitions/TagIdentifier"
          },
          "maxItems": 10,
          "examples": ["stablecoin", "compound"]
        },
        "extensions": {
          "type": "object",
          "description": "An object containing any arbitrary or vendor-specific token metadata",
          "propertyNames": {
            "$ref": "#/definitions/ExtensionIdentifier"
          },
          "additionalProperties": {
            "$ref": "#/definitions/ExtensionValue"
          },
          "maxProperties": 10,
          "examples": [
            {
              "color": "#000000",
              "is_verified_by_me": true
            }
          ]
        }
      },
      "required": ["chainId", "address", "decimals", "name", "symbol"]
    }
  },
  "type": "object",
  "additionalProperties": false,
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the protection market list",
      "minLength": 1,
      "maxLength": 255,
      "pattern": "^[\\w ]+$",
      "examples": ["My Protection Market List"]
    },
    "timestamp": {
      "type": "string",
      "format": "date-time",
      "description": "The timestamp of this list version; i.e. when this immutable version of the list was created"
    },
    "version": {
      "$ref": "#/definitions/Version"
    },
    "markets": {
      "type": "array",
      "description": "The list of markets included in the list",
      "items": {
        "$ref": "#/definitions/ProtectionMarketInfo"
      },
      "minItems": 1,
      "maxItems": 10000
    },
    "keywords": {
      "type": "array",
      "description": "Keywords associated with the contents of the list; may be used in list discoverability",
      "items": {
        "type": "string",
        "description": "A keyword to describe the contents of the list",
        "minLength": 1,
        "maxLength": 20,
        "pattern": "^[\\w ]+$",
        "examples": ["yearn", "lending"]
      },
      "maxItems": 20,
      "uniqueItems": true
    },
    "tags": {
      "type": "object",
      "description": "A mapping of tag identifiers to their name and description",
      "propertyNames": {
        "$ref": "#/definitions/TagIdentifier"
      },
      "additionalProperties": {
        "$ref": "#/definitions/TagDefinition"
      },
      "maxProperties": 20,
      "examples": [
        {
          "stablecoin": {
            "name": "Stablecoin",
            "description": "A token with value pegged to another asset"
          }
        }
      ]
    },
    "logoURI": {
      "type": "string",
      "description": "A URI for the logo of the protection market list; prefer SVG or PNG of size 256x256",
      "format": "uri",
      "examples": ["ipfs://QmXfzKRvjZz3u5JRgC4v5mGVbm9ahrUiB4DgzHBsnWbTMM"]
    }
  },
  "required": ["name", "timestamp", "version", "markets"]
}
