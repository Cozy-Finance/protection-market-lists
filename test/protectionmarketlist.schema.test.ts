import Ajv from 'ajv';
import { schema } from '../src';
import exampleList from './schema/example.protectionmarketlist.json';
import exampleNameSymbolSpecialCharacters from './schema/example-name-symbol-special-characters.protectionmarketlist.json';
import bigExampleList from './schema/bigexample.protectionmarketlist.json';
import exampleListMinimum from './schema/exampleminimum.protectionmarketlist.json';
import emptyList from './schema/empty.protectionmarketlist.json';
import bigWords from './schema/bigwords.protectionmarketlist.json';
import invalidTokenAddress1 from './schema/invalidtokenaddress.1.protectionmarketlist.json';
import invalidTokenAddress2 from './schema/invalidtokenaddress.2.protectionmarketlist.json';
import invalidTriggerAddress from './schema/invalidtriggeraddress.protectionmarketlist.json';
import invalidTimestamp from './schema/invalidtimestamp.protectionmarketlist.json';
import invalidLogoURI1 from './schema/invalidlogouri.1.protectionmarketlist.json';
import invalidLogoURI2 from './schema/invalidlogouri.2.protectionmarketlist.json';
import invalidVersion1 from './schema/invalidversion.1.protectionmarketlist.json';
import invalidVersion2 from './schema/invalidversion.2.protectionmarketlist.json';
import invalidVersion3 from './schema/invalidversion.3.protectionmarketlist.json';
import invalidDecimals1 from './schema/invaliddecimals.1.protectionmarketlist.json';
import invalidNumTags from './schema/invalidNumTags.protectionmarketlist.json';
import invalidDecimals2 from './schema/invaliddecimals.2.protectionmarketlist.json';
import invalidDecimals3 from './schema/invaliddecimals.3.protectionmarketlist.json';
import invalidDecimals4 from './schema/invaliddecimals.4.protectionmarketlist.json';
import extensionsValid from './schema/extensions-valid.protectionmarketlist.json';
import extensionsInvalid from './schema/extensions-invalid.protectionmarketlist.json';
import addFormats from 'ajv-formats';

const ajv = new Ajv({ allErrors: true });
addFormats(ajv);
const validator = ajv.compile(schema);

describe('schema', () => {
  it('is valid', () => {
    expect(ajv.validateSchema(schema)).toEqual(true);
  });

  function checkSchema(schema: any, valid: boolean): void {
    const isValid = validator(schema);
    expect(validator.errors).toMatchSnapshot();
    expect(isValid).toEqual(valid);
  }

  it('works for example schema', () => {
    checkSchema(exampleList, true);
  });

  it('works for special characters schema', () => {
    checkSchema(exampleNameSymbolSpecialCharacters, true);
  });

  it('works for big example schema', () => {
    checkSchema(bigExampleList, true);
  });

  it('minimum example schema', () => {
    checkSchema(exampleListMinimum, true);
  });

  it('requires name, timestamp, version, tokens', () => {
    checkSchema({}, false);
  });

  it('empty list fails', () => {
    checkSchema(emptyList, false);
  });

  it('fails with big names', () => {
    checkSchema(bigWords, false);
  });

  it('checks token address', () => {
    checkSchema(invalidTokenAddress1, false);
    checkSchema(invalidTokenAddress2, false);
  });

  it('checks trigger address', () => {
    checkSchema(invalidTriggerAddress, false);
  });

  it('invalid timestamp', () => {
    checkSchema(invalidTimestamp, false);
  });

  it('invalid logo URI', () => {
    checkSchema(invalidLogoURI1, false);
    checkSchema(invalidLogoURI2, false);
  });

  it('invalid decimals', () => {
    checkSchema(invalidDecimals1, false);
    checkSchema(invalidDecimals2, false);
    checkSchema(invalidDecimals3, false);
    checkSchema(invalidDecimals4, false);
  });

  it('invalid number of tags on token', () => {
    checkSchema(invalidNumTags, false);
  });

  it('checks version', () => {
    checkSchema(invalidVersion1, false);
    checkSchema(invalidVersion2, false);
    checkSchema(invalidVersion3, false);
  });

  it('checks extensions', () => {
    checkSchema(extensionsValid, true);
    checkSchema(extensionsInvalid, false);
  });

  it('allows up to 10k tokens', () => {
    const exampleListWith10kTokens = {
      ...exampleList,
      markets: [...Array(10000)].map(() => exampleList.markets[0]),
    };
    checkSchema(exampleListWith10kTokens, true);
  });

  it('fails with 10001 tokens', () => {
    const exampleListWith10kTokensPlusOne = {
      ...exampleList,
      markets: [...Array(10001)].map(() => exampleList.markets[0]),
    };
    checkSchema(exampleListWith10kTokensPlusOne, false);
  });
});
