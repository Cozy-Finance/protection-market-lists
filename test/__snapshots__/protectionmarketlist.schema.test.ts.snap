// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema allows up to 10k tokens 1`] = `null`;

exports[`schema checks extensions 1`] = `null`;

exports[`schema checks extensions 2`] = `
Array [
  Object {
    "instancePath": "/markets/0/extensions/abc",
    "keyword": "type",
    "message": "must be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/0/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/abc",
    "keyword": "type",
    "message": "must be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/1/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/abc",
    "keyword": "type",
    "message": "must be number",
    "params": Object {
      "type": "number",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/2/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/abc",
    "keyword": "type",
    "message": "must be null",
    "params": Object {
      "type": "null",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/3/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/abc",
    "keyword": "anyOf",
    "message": "must match a schema in anyOf",
    "params": Object {},
    "schemaPath": "#/definitions/ExtensionValue/anyOf",
  },
  Object {
    "instancePath": "/markets/0/extensions/def",
    "keyword": "type",
    "message": "must be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/0/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/def",
    "keyword": "type",
    "message": "must be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/1/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/def",
    "keyword": "type",
    "message": "must be number",
    "params": Object {
      "type": "number",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/2/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/def",
    "keyword": "type",
    "message": "must be null",
    "params": Object {
      "type": "null",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/3/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/def",
    "keyword": "anyOf",
    "message": "must match a schema in anyOf",
    "params": Object {},
    "schemaPath": "#/definitions/ExtensionValue/anyOf",
  },
  Object {
    "instancePath": "/markets/0/extensions/isBoolean",
    "keyword": "type",
    "message": "must be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/0/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/isBoolean",
    "keyword": "type",
    "message": "must be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/1/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/isBoolean",
    "keyword": "type",
    "message": "must be number",
    "params": Object {
      "type": "number",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/2/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/isBoolean",
    "keyword": "type",
    "message": "must be null",
    "params": Object {
      "type": "null",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/3/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/isBoolean",
    "keyword": "anyOf",
    "message": "must match a schema in anyOf",
    "params": Object {},
    "schemaPath": "#/definitions/ExtensionValue/anyOf",
  },
  Object {
    "instancePath": "/markets/0/extensions/isNull",
    "keyword": "type",
    "message": "must be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/0/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/isNull",
    "keyword": "type",
    "message": "must be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/1/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/isNull",
    "keyword": "type",
    "message": "must be number",
    "params": Object {
      "type": "number",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/2/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/isNull",
    "keyword": "type",
    "message": "must be null",
    "params": Object {
      "type": "null",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/3/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/isNull",
    "keyword": "anyOf",
    "message": "must match a schema in anyOf",
    "params": Object {},
    "schemaPath": "#/definitions/ExtensionValue/anyOf",
  },
  Object {
    "instancePath": "/markets/0/extensions/tooLong",
    "keyword": "maxLength",
    "message": "must NOT have more than 42 characters",
    "params": Object {
      "limit": 42,
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/0/maxLength",
  },
  Object {
    "instancePath": "/markets/0/extensions/tooLong",
    "keyword": "type",
    "message": "must be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/1/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/tooLong",
    "keyword": "type",
    "message": "must be number",
    "params": Object {
      "type": "number",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/2/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/tooLong",
    "keyword": "type",
    "message": "must be null",
    "params": Object {
      "type": "null",
    },
    "schemaPath": "#/definitions/ExtensionValue/anyOf/3/type",
  },
  Object {
    "instancePath": "/markets/0/extensions/tooLong",
    "keyword": "anyOf",
    "message": "must match a schema in anyOf",
    "params": Object {},
    "schemaPath": "#/definitions/ExtensionValue/anyOf",
  },
]
`;

exports[`schema checks token address 1`] = `
Array [
  Object {
    "instancePath": "/markets/0/address",
    "keyword": "pattern",
    "message": "must match pattern \\"^0x[a-fA-F0-9]{40}$\\"",
    "params": Object {
      "pattern": "^0x[a-fA-F0-9]{40}$",
    },
    "schemaPath": "#/properties/address/pattern",
  },
]
`;

exports[`schema checks token address 2`] = `
Array [
  Object {
    "instancePath": "/tokens/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/address",
    "keyword": "pattern",
    "message": "must match pattern \\"^0x[a-fA-F0-9]{40}$\\"",
    "params": Object {
      "pattern": "^0x[a-fA-F0-9]{40}$",
    },
    "schemaPath": "#/properties/address/pattern",
  },
]
`;

exports[`schema checks trigger address 1`] = `
Array [
  Object {
    "instancePath": "/markets/0/trigger",
    "keyword": "pattern",
    "message": "must match pattern \\"^0x[a-fA-F0-9]{40}$\\"",
    "params": Object {
      "pattern": "^0x[a-fA-F0-9]{40}$",
    },
    "schemaPath": "#/properties/trigger/pattern",
  },
]
`;

exports[`schema checks version 1`] = `
Array [
  Object {
    "instancePath": "/version/patch",
    "keyword": "minimum",
    "message": "must be >= 0",
    "params": Object {
      "comparison": ">=",
      "limit": 0,
    },
    "schemaPath": "#/definitions/Version/properties/patch/minimum",
  },
]
`;

exports[`schema checks version 2`] = `
Array [
  Object {
    "instancePath": "/version/minor",
    "keyword": "minimum",
    "message": "must be >= 0",
    "params": Object {
      "comparison": ">=",
      "limit": 0,
    },
    "schemaPath": "#/definitions/Version/properties/minor/minimum",
  },
]
`;

exports[`schema checks version 3`] = `
Array [
  Object {
    "instancePath": "/version/major",
    "keyword": "minimum",
    "message": "must be >= 0",
    "params": Object {
      "comparison": ">=",
      "limit": 0,
    },
    "schemaPath": "#/definitions/Version/properties/major/minimum",
  },
]
`;

exports[`schema empty list fails 1`] = `
Array [
  Object {
    "instancePath": "/markets",
    "keyword": "minItems",
    "message": "must NOT have fewer than 1 items",
    "params": Object {
      "limit": 1,
    },
    "schemaPath": "#/properties/markets/minItems",
  },
]
`;

exports[`schema fails with 10001 tokens 1`] = `
Array [
  Object {
    "instancePath": "/markets",
    "keyword": "maxItems",
    "message": "must NOT have more than 10000 items",
    "params": Object {
      "limit": 10000,
    },
    "schemaPath": "#/properties/markets/maxItems",
  },
]
`;

exports[`schema fails with big names 1`] = `
Array [
  Object {
    "instancePath": "/markets/0/name",
    "keyword": "maxLength",
    "message": "must NOT have more than 100 characters",
    "params": Object {
      "limit": 100,
    },
    "schemaPath": "#/properties/name/maxLength",
  },
]
`;

exports[`schema invalid decimals 1`] = `
Array [
  Object {
    "instancePath": "/markets/0/decimals",
    "keyword": "maximum",
    "message": "must be <= 255",
    "params": Object {
      "comparison": "<=",
      "limit": 255,
    },
    "schemaPath": "#/properties/decimals/maximum",
  },
]
`;

exports[`schema invalid decimals 2`] = `
Array [
  Object {
    "instancePath": "/markets/0/decimals",
    "keyword": "minimum",
    "message": "must be >= 0",
    "params": Object {
      "comparison": ">=",
      "limit": 0,
    },
    "schemaPath": "#/properties/decimals/minimum",
  },
]
`;

exports[`schema invalid decimals 3`] = `
Array [
  Object {
    "instancePath": "/tokens/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/decimals",
    "keyword": "maximum",
    "message": "must be <= 255",
    "params": Object {
      "comparison": "<=",
      "limit": 255,
    },
    "schemaPath": "#/properties/decimals/maximum",
  },
]
`;

exports[`schema invalid decimals 4`] = `
Array [
  Object {
    "instancePath": "/tokens/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/decimals",
    "keyword": "minimum",
    "message": "must be >= 0",
    "params": Object {
      "comparison": ">=",
      "limit": 0,
    },
    "schemaPath": "#/properties/decimals/minimum",
  },
]
`;

exports[`schema invalid logo URI 1`] = `
Array [
  Object {
    "instancePath": "/markets/0/logoURI",
    "keyword": "format",
    "message": "must match format \\"uri\\"",
    "params": Object {
      "format": "uri",
    },
    "schemaPath": "#/properties/logoURI/format",
  },
]
`;

exports[`schema invalid logo URI 2`] = `
Array [
  Object {
    "instancePath": "/tokens/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/logoURI",
    "keyword": "format",
    "message": "must match format \\"uri\\"",
    "params": Object {
      "format": "uri",
    },
    "schemaPath": "#/properties/logoURI/format",
  },
]
`;

exports[`schema invalid number of tags on token 1`] = `
Array [
  Object {
    "instancePath": "/markets/0/tags",
    "keyword": "maxItems",
    "message": "must NOT have more than 10 items",
    "params": Object {
      "limit": 10,
    },
    "schemaPath": "#/properties/tags/maxItems",
  },
]
`;

exports[`schema invalid timestamp 1`] = `
Array [
  Object {
    "instancePath": "/timestamp",
    "keyword": "format",
    "message": "must match format \\"date-time\\"",
    "params": Object {
      "format": "date-time",
    },
    "schemaPath": "#/properties/timestamp/format",
  },
]
`;

exports[`schema minimum example schema 1`] = `null`;

exports[`schema requires name, timestamp, version, tokens 1`] = `
Array [
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'name'",
    "params": Object {
      "missingProperty": "name",
    },
    "schemaPath": "#/required",
  },
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'timestamp'",
    "params": Object {
      "missingProperty": "timestamp",
    },
    "schemaPath": "#/required",
  },
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'version'",
    "params": Object {
      "missingProperty": "version",
    },
    "schemaPath": "#/required",
  },
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'markets'",
    "params": Object {
      "missingProperty": "markets",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`schema works for big example schema 1`] = `null`;

exports[`schema works for example schema 1`] = `null`;

exports[`schema works for special characters schema 1`] = `null`;
